### PR TITLE
Cache - base filename on path with options and save with item for clearing on update

### PR DESCRIPTION
**Pull request by DaveTBlake**
Follow up to #16805 and needed to fix video library watched state issues #15251 and #16357 
Building cache filename using just path without options string has caused unintended consequnces for window navigation for items from plugins.  Also while it did solve cache identification and clearing issues for library item lists, it didn't do so for plugin items. Plugin lists all have the same base path, the option string is needed to differentiate between them!

The underlying problem that this solves (and that both #16805 and prior to that #14810 attempted to solve) was observed as follows: on playback of a video finishing, if the list of videos is subsequently read from cache (because it was slow to load as MySQL server over LAN or lots of items etc. ) then the watched state and playcount of that item is not updated on the GUI.

Looking at the  paths used to form cache file name reveals the source of this wrong behaviour. Say you have navigated to a list of episodes and are playing an episode. 
List path is `videodb://tvshows/titles/5/3?tvshowid=5&xsp...`
Item path is `videodb://tvshows/titles/5/3/68?season=3&tvshowid=5&xsp...`
A cache file is created for the list using the list path. When playback completes an update item message is sent and `CGUIMediaWindow` attempts to clear the cache for the list with this item. To get the name of the cache file it tries to recover the list path from the item path by trucating it to `videodb://tvshows/titles/5/3/` removing last parameter and options. This is not the list path, and the wrong cache filename is used and the cache for the list is not cleared.

Or say you were using the Netflix plugin
List path `plugin://plugin.video.netflix/?action=video_lists&profile_id=ZHX...`  gets cached
Item path ` plugin://plugin.video.netflix/?action=video_list&type=netflixOriginals&video_list_id=318...`
On item update the truncate approach to getting list path from item path would give incorrectly `plugin://plugin.video.netflix/` and again the cache is not cleared. User does not see updated watched state or playcount etc.

Cache filename needs to be based on path with options, and there is no easy way to get list path from the item path. Hence my solution is to store the cache filename as an item property for every item in the list, that way on item update the correct cache file can be found and removed.

In summary:
Base cache filename on itemlist path with options.
Save cache filename with each item in list  and use on item update to remove correct cache file.

@peak3d this should solve the plugin issues (original and new) @spiff for a saneity check

### CODE REVIEW

File: xbmc/FileItem.cpp [hunk 1]
```diff
@@ -2920,8 +2920,17 @@ bool CFileItemList::Save(int windowID)
   CLog::Log(LOGDEBUG,"Saving fileitems [%s]", CURL::GetRedacted(GetPath()).c_str());
 
   CFile file;
-  if (file.OpenForWrite(GetDiscFileCache(windowID), true)) // overwrite always
+  std::string cachefile = GetDiscFileCache(windowID);
+  if (file.OpenForWrite(cachefile, true)) // overwrite always
   {
+    // Before caching save simplified cache file name in every item so the cache file can be
```
Comment (Rechi) [inline xbmc/FileItem.cpp]: remove the white space at the end of the line

File: xbmc/FileItem.cpp [hunk 2]
```diff
@@ -2920,8 +2920,17 @@ bool CFileItemList::Save(int windowID)
   CLog::Log(LOGDEBUG,"Saving fileitems [%s]", CURL::GetRedacted(GetPath()).c_str());
 
   CFile file;
-  if (file.OpenForWrite(GetDiscFileCache(windowID), true)) // overwrite always
+  std::string cachefile = GetDiscFileCache(windowID);
+  if (file.OpenForWrite(cachefile, true)) // overwrite always
   {
+    // Before caching save simplified cache file name in every item so the cache file can be 
+    // identifed and removed if the item is updated. List path and options (used for file
```
Comment (Rechi) [inline xbmc/FileItem.cpp]: remove the white space at the end of the line

File: xbmc/FileItem.cpp [hunk 3]
```diff
@@ -2920,8 +2920,17 @@ bool CFileItemList::Save(int windowID)
...
-  if (file.OpenForWrite(GetDiscFileCache(windowID), true)) // overwrite always
+  std::string cachefile = GetDiscFileCache(windowID);
+  if (file.OpenForWrite(cachefile, true)) // overwrite always
   {
+    // Before caching save simplified cache file name in every item so the cache file can be 
+    // identifed and removed if the item is updated. List path and options (used for file 
+    // name when list cached) can not be accurately derived from item path.
+    StringUtils::Replace(cachefile, "special://temp/archive_cache/", "");
+    StringUtils::Replace(cachefile, ".fi", "");
+    for (auto item :  m_items)
```
Comment (Rechi) [inline xbmc/FileItem.cpp]: > Conventional operators have to be surrounded by one space on each side.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#34-whitespace

**Comment by DaveTBlake** [comment]
White space corrections done.

**Comment by peak3d** [comment]
:+1 works as expected, thx @DaveTBlake

**Comment by DaveTBlake** [comment]
jenkins build this please

**Comment by DaveTBlake** [comment]
Have also had positive test feedback from users of the v18 test build for this fix backported, see https://github.com/xbmc/xbmc/issues/16357

**Comment by DaveTBlake** [comment]
jenkins build this please

**Comment by DaveTBlake** [comment]
jenkins build this please

**Comment by DaveTBlake** [comment]
jenkins now giving build error on OSX https://jenkins.kodi.tv/job/OSX-64/15697/console
I don't see how this relates to this PR, anyone advise?

**Comment by asavah** [comment]
Did extensive "youtubing" with master build with this PR included, mostly large playlists, works as described.
Thx.

**Comment by DaveTBlake** [comment]
jenkins build this please

**Comment by DaveTBlake** [comment]
UWP build faliure is Py3 related, not this PR.
